File: Civitas_Ruby/lib/vista_textual.rb
#encoding:utf-8
require_relative 'operaciones_juego.rb'
require_relative 'civitas_juego.rb'
require 'io/console'
require_relative 'jugador.rb'
require_relative 'titulo_propiedad.rb'
require_relative 'civitas_juego.rb'
require_relative 'operacion_inmobiliaria.rb'
require_relative 'gestiones_inmobiliarias.rb'
require_relative 'diario.rb'
require_relative 'salidas_carcel.rb'
require_relative 'respuestas.rb'


module Civitas

  class Vista_textual
    attr_reader :juegoModel, :iGestion, :iPropiedad
    def initialize()
      @iGestion = -1
      @iPropiedad = -1
      @juegoModel
      @separador = "====================="
    end

    def mostrar_estado(estado)
      puts estado
    end

    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end



    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
      return opcion
    end

    
    def comprar()
      respuestas = [Respuestas::NO, Respuestas::SI]
      respuestas_s = ["no", "si"]
      opcion = menu("¿Quieres comprar la calle?", respuestas_s)
      
      return respuestas[opcion]
    end

    def gestionar
      
      respuestas = [GestionesInmobiliarias::VENDER, GestionesInmobiliarias::HIPOTECAR, GestionesInmobiliarias::CANCELAR_HIPOTECA, GestionesInmobiliarias::CONSTRUIR_CASA, GestionesInmobiliarias::CONSTRUIR_HOTEL, GestionesInmobiliarias::TERMINAR]
      respuestas_s = ["vender", "hipotecar", "cancelar_hipoteca", "construir_casa", "construir_hotel", "terminar"]
      opcion = menu("Elige la gestión inmobiliria que quieres realizar",respuestas_s)
      @iGestion = opcion
      
   
      
      if(@iGestion != 5)
        propiedades_name = []
        propiedades_titulo = @juegoModel.getJugadorActual().propiedades
        for i in 0..propiedades_titulo.size()-1
          propiedades_name.push(propiedades_titulo[i].nombre)
        end
        
        opcion = menu("¿Sobre qué propiedad?", propiedades_name)
        @iPropiedad = opcion
      end
      
      
    end


    def mostrarSiguienteOperacion(operacion)
      puts operacion
    end

    def mostrarEventos
      while Diario.instance.eventos_pendientes()
        puts Diario.instance.leer_evento()
      end
    end

    def setCivitasJuego(civitas)
         @juegoModel=civitas
         self.actualizarVista
    end

    def actualizarVista
      puts @juegoModel.getJugadorActual().to_s()
      puts @juegoModel.getCasillaActual().to_s()
    end
    
    def salirCarcel
      formas_salir = [SalidasCarcel::PAGANDO, SalidasCarcel::TIRANDO]
      formas_salir_s = ["pagando","tirando"]
      opcion = menu("Elige la forma para salir de la carcel", formas_salir_s)
     
      return formas_salir[opcion]
    end

    
  end

end

File: Civitas_TUI/src/JuegoTexto/VistaTextual.java
package JuegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.Jugador;
import civitas.Respuestas;
import civitas.TituloPropiedad;

public class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  public VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar() {
      int opcion = menu("¿Quieres comprar la calle?", new ArrayList<> (Arrays.asList("No","Si")));

      return (Respuestas.values()[opcion]);
  }

  void gestionar () {
      int opcion = menu("Elige la gestión inmobiliaria que quieres realizar", new ArrayList<> (Arrays.asList("Vender", "Hipotecar", "Cancelar Hipoteca", "Construir Casa", "Construir Hotel", "Terminar")));
      iGestion = opcion;
      if(iGestion != 5){
        ArrayList<String> propiedades_name = new ArrayList<>();
        ArrayList<TituloPropiedad> propiedades_titulo = new ArrayList<>();
        propiedades_titulo = juegoModel.getJugadorActual().getPropiedades();
        for(int i = 0; i < propiedades_titulo.size(); i++){
            propiedades_name.add(propiedades_titulo.get(i).getNombre());
        }
        
        opcion = menu("¿Sobre qué propiedad?", propiedades_name);
        
        iPropiedad = opcion;
      }
  }
  
  public int getGestion(){
      return iGestion;
  }
  
  public int getPropiedad(){
      return iPropiedad;
  }
    

  void mostrarSiguienteOperacion(OperacionesJuego operacion) {
      System.out.println(operacion);
  }


  void mostrarEventos() {
      while(Diario.getInstance().eventosPendientes())
          System.out.println(Diario.getInstance().leerEvento());
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        this.actualizarVista();
    }
  
  void actualizarVista(){
      System.out.println(juegoModel.getJugadorActual().toString());
      System.out.println(juegoModel.getCasillaActual().toString());
  } 
}
